pkg/provider/cloud/vsphere/tags.go

```diff
@@ -29,78 +29,158 @@ import (
+			cluster.Status.ProviderStatus.VSphere = &kubermaticv1.VSphereStatus{}
+		}
+		cluster.Status.ProviderStatus.VSphere.ManagedTagIDs = sets.List(managedTags)
+		return cluster, managedTags, nil
+	}
+	updatedCluster, err := update(ctx, cluster.Name, func(cluster *kubermaticv1.Cluster) {
 		if !kuberneteshelper.HasFinalizer(cluster, tagCleanupFinalizer) {
 			kuberneteshelper.AddFinalizer(cluster, tagCleanupFinalizer)
 		}
 	})
-	if err != nil {
-		return nil, fmt.Errorf("failed to add finalizer %s on vsphere cluster object: %w", tagCleanupFinalizer, err)
-	}
-
-	return cluster, nil
+	return updatedCluster, managedTags, err
 }
 
-func syncCreatedClusterTags(ctx context.Context, restSession *RESTSession, cluster *kubermaticv1.Cluster) error {
+func syncCreatedClusterTags(ctx context.Context, restSession *RESTSession, cluster *kubermaticv1.Cluster) (sets.Set[string], error) {
 	tagManager := vapitags.NewManager(restSession.Client)
 	categoryTags, err := tagManager.GetTagsForCategory(ctx, cluster.Spec.Cloud.VSphere.Tags.CategoryID)
 	if err != nil {
-		return fmt.Errorf("failed to get tag category %s: %w", cluster.Spec.Cloud.VSphere.Tags.CategoryID, err)
+		return nil, fmt.Errorf("failed to get tag category %s: %w", cluster.Spec.Cloud.VSphere.Tags.CategoryID, err)
 	}
 
-	for _, vsphereTag := range cluster.Spec.Cloud.VSphere.Tags.Tags {
-		if filterTag(categoryTags, vsphereTag) == "" {
-			_, err := createTag(ctx, restSession, cluster.Spec.Cloud.VSphere.Tags.CategoryID, vsphereTag)
-			if err != nil {
-				return fmt.Errorf("failed to create tag %s against category %s: %w", vsphereTag, cluster.Spec.Cloud.VSphere.Tags.CategoryID, err)
+	managedTags := getManagedTags(cluster)
+	specTags := sets.NewString(cluster.Spec.Cloud.VSphere.Tags.Tags...)
+
+	// Ensure all tags in the spec are present in vSphere and tracked in the annotation.
+	for _, tagName := range specTags.List() {
+		// Check if the tag from the spec already exists in vSphere for this category.
+		if tagID := filterTag(categoryTags, tagName); tagID != "" {
+			// If it exists, ensure it's in our managed list.
+			if !managedTags.Has(tagID) {
+				managedTags.Insert(tagID)
 			}
+			continue
 		}
+
+		// If the tag does not exist, create it.
+		tagID, err := createTag(ctx, restSession, cluster.Spec.Cloud.VSphere.Tags.CategoryID, tagName)
+		if err != nil {
+			return nil, fmt.Errorf("failed to create tag %s against category %s: %w", tagName, cluster.Spec.Cloud.VSphere.Tags.CategoryID, err)
+		}
+		managedTags.Insert(tagID)
 	}
 
-	return nil
+	return managedTags, nil
 }
 
-func syncDeletedClusterTags(ctx context.Context, restSession *RESTSession, cluster *kubermaticv1.Cluster) error {
+func syncDeletedClusterTags(ctx context.Context, restSession *RESTSession, cluster *kubermaticv1.Cluster, managedTags sets.Set[string]) (sets.Set[string], error) {
+	// If there are no managed tags, there's nothing to delete.
+	if managedTags.Len() == 0 {
+		return managedTags, nil
+	}
+
```

Comment:
This code is unfortunately quite unreadable and was difficult to review, would be great if you can improve the code quality a bit.